src/backend/distributed/operations/worker_shard_copy.c

```diff
@@ -232,6 +234,27 @@ ShardCopyDestReceiverReceive(TupleTableSlot *slot, DestReceiver *dest)
 	return true;
 }
 
+static void 
+SetupReplicationOrigin(RepOriginId nodeId) {
+	RepOriginId originid = InvalidRepOriginId;
+	XLogRecPtr	origin_startpos = InvalidXLogRecPtr;
+	//Check if there is a replication origin session already active.
+	if (replorigin_session_origin == InvalidRepOriginId) {
+		//Lookup the replication origin and create it if it does not exist.
+		char originname[NAMEDATALEN];
+		snprintf(originname, sizeof(originname), "pg_%u",nodeId);
+		originid = replorigin_by_name(originname, true);
+		if (originid == InvalidRepOriginId) {
+			originid = replorigin_create(originname);
+		}
+		//Setup the replication origin session.
+		replorigin_session_setup(originid);
```

Comment:
It seems this sets up a replication origin in the session that does reading+sending (pushing) on the source, but I think we should need to set it up in the receiving+writing session using the UDFs?

See https://www.postgresql.org/docs/current/functions-admin.html#PG-REPLICATION-ORIGIN-SESSION-SETUP

Reply:
(what a nice surprise to see Marco's comment :) )

I was also confused at first, and still not looked deeper yet. But my current understanding is that `filter_by_origin_cb ` is called on the target/receiving part.

So, we call `replorigin_session_setup ` on the source, then the origin is written to the WAL and `filter_by_origin_cb` filters that on the target.

But, yes, needs a closer look I agree, some relevant discussion [here](https://github.com/citusdata/citus/pull/6453#discussion_r1006534331)

Reply:
> It seems this sets up a replication origin in the session that does reading+sending (pushing) on the source, but I think we need to set it up in the receiving+writing session using the UDFs?

Alright, we discussed this @rajeshkt78 and agreed on your statement, @rajeshkt78 is going to reflect this to the PR with some description to the PR as well

Reply:
> With PG [15+](https://github.com/postgres/postgres/commit/366283961ac0ed6d89014444c6090f3fd02fce0a), we can configure whether a tuple that has a replication origin should be replicated again or not. If we do, CREATE SUBSCRIPTION sub1 CONNECTION '.. PUBLICATION pub1 WITH (**origin = none**);, we can control the behavior.

Sorry, Marco realized that this is PG 16 not PG 15, so we cannot rely on this. But, we could probably follow a similar approach

Reply:
I have changed the implementation to setup replication origin in remote nodes instead of source nodes.